ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java

```diff
@@ -1411,6 +1411,108 @@ public void testRouterMetadataCacheSimpleBlob() throws Exception {
+   * @throws Exception
+   */
+  @Test
+  public void testRouterMetadataCacheEviction() throws Exception {
+    try {
+      int NUM_PUT_COMPOSITE_BLOBS = 3;
+      int NUM_GET_COMPOSITE_BLOBS = NUM_PUT_COMPOSITE_BLOBS * 5;
+      int TEST_MAX_NUM_METADATA_CACHE_ENTRIES = 1;
+      int numMisses = NUM_GET_COMPOSITE_BLOBS;
+      int numInserts = NUM_GET_COMPOSITE_BLOBS;
+      Properties properties = getNonBlockingRouterProperties("DC1");
+      properties.setProperty("router.blob.metadata.cache.enabled", Boolean.toString(true));
+      properties.setProperty("router.smallest.blob.for.metadata.cache", Long.toString(0));
+      // Set the size of cache to 1 entry which means the cache can store at max of one entry at any time
+      properties.setProperty(RouterConfig.ROUTER_MAX_NUM_METADATA_CACHE_ENTRIES, Integer.toString(TEST_MAX_NUM_METADATA_CACHE_ENTRIES));
+      setRouterWithMetadataCache(properties, new AmbryCacheMetrics(0, numMisses, numInserts,0));
+      AmbryCacheWithMetrics ambryCacheWithMetrics = (AmbryCacheWithMetrics) router.getBlobMetadataCache();
+      assertNotNull("Metadata cache must be created", ambryCacheWithMetrics);
+      assertEquals(String.format("Expected %s to be %s but found %s",
+          RouterConfig.ROUTER_MAX_NUM_METADATA_CACHE_ENTRIES, TEST_MAX_NUM_METADATA_CACHE_ENTRIES, ambryCacheWithMetrics.getMaxNumCacheEntries()),
+          TEST_MAX_NUM_METADATA_CACHE_ENTRIES, ambryCacheWithMetrics.getMaxNumCacheEntries());
+
+      ArrayList<String> blobIds = new ArrayList<>();
+      for(int i = 0; i < NUM_PUT_COMPOSITE_BLOBS; ++i) {
+        // put few composite blob and test it is absent in the metadata cache
+        setOperationParams(2 * PUT_CONTENT_SIZE, TTL_SECS);
+        String blobId =
+            router.putBlob(putBlobProperties, putUserMetadata, putChannel, new PutBlobOptionsBuilder().build())
+                .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        blobIds.add(blobId);
+      }
+
+      for(int i = 0; i < NUM_GET_COMPOSITE_BLOBS; ++i) {
+        // cache miss and populate cache
+        // get the one byte from composite blob, and it's metadata must be present in cache
+        ByteRange range = ByteRanges.fromOffsetRange(1, 2);
+        String blobId = blobIds.get(i % NUM_PUT_COMPOSITE_BLOBS);
+        router.getBlob(blobId, new GetBlobOptionsBuilder().range(range).build()).get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+      }
+
+      assertTrue(String.format("Must encounter %s cache misses", numMisses), ambryCacheWithMetrics.getCacheMissCountDown().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertTrue(String.format("Must encounter %s cache inserts", numInserts), ambryCacheWithMetrics.getPutObjectCountDown().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+    } finally {
```

Comment:
I guess this is a mocked router.  It is created once but closed multiple times (one in each test).  If it's a mock, don't have to do anything about it.

Reply:
for each test, the entire test class is instantiated, router created, test run and then router closed.